### ISSUE TITLE
Subcycling doesn't work when bundling processes together

### ISSUE THREAD

**Issue by AaronDonahue** [E3SM-Project/scream#1467]
The ability to subcycle processes or groups of processes was added in #1371 , which also has instructions on how to set up subcycling for a set of tightly coupled processes.  Unfortunately this doesn't appear to work.

The issue appears to be because we set one of the main processes to be a Process Group, which is not registered as a viable process.  Like SPA, SHOC, P3, RRMTGP, CLD are for example.  So the AD throws an error.

A simple re-creator is to adjust one of the unit tests, in this case I am adjusting the `shoc_cld_spa_p3_rrtmgp` test.

```
  {Unknown expression after the reported line}
due to unexpected exception with message:
  /g/g17/donahue5/Code/e3sm/scream/scream/externals/ekat/src/ekat/util/
  ekat_factory.hpp:142: FAIL:
  it!=m_register.end()
  [N4ekat7FactoryIN6scream17AtmosphereProcessENS_21CaseInsensitiveStringESt10s-
  hared_ptrIS2_EJRKNS_4CommERKNS_13ParameterListEEEE] Error!
          The key 'Group' is not associated to any registered product.
          The list of registered product is: CldFraction, p3, RRTMGP, SHOC, SPA
          Did you forget to register it?

===============================================================================
test cases: 1 | 1 failed
assertions: 2 | 1 passed | 1 failed

[donahue5@lassen17:shoc_cld_spa_p3_rrtmgp]$ diff input.yaml input.yaml.old 
13c13
<   Number of Entries: 4
---
>   Number of Entries: 5
27,38c27,33
<     Process Name: Group
<     Number of Entries: 2
<     Schedule Type: Sequential
<     Number of Subcycles: 3
<     Process 0:
<       Process Name: P3
<       Grid: Point Grid
<     Process 1:
<       Process Name: RRTMGP
<       Grid: Point Grid
<       active_gases: ["h2o", "co2", "o3", "n2o", "co" , "ch4", "o2", "n2"]
<       Orbital Year: 1990
---
>     Process Name: P3
>     Grid: Point Grid
>   Process 4:
>     Process Name: RRTMGP
>     Grid: Point Grid
>     active_gases: ["h2o", "co2", "o3", "n2o", "co" , "ch4", "o2", "n2"]
>     Orbital Year: 1990

```

**Comment by bartgol** [comment]
@AaronDonahue This is not an infrastructure error though. You simply have to register the process, like you do for dyn and phys, in the test cpp file, before creating the AD (that's . Admittedly, we could find a way to register that automatically, but I can't think of any clean way, other than adding it to both `register_physics.hpp` and `register_dynamics.hpp`.

So to be clear, subcycling _does_ work for a "bundle" of processes. It's the process factory that is failing to create the group, since nobody register the corresponding creator in the factory.

**Comment by AaronDonahue** [comment]
right, I think it is a straightforward fix.  But right now, subcycling doesn't work in any of our tests because group isn't registered.  But it's not so simple as adding it to the list of registered processes.  For example, the following results in a compile time error:

```
diff --git a/components/scream/src/physics/register_physics.hpp b/components/scream/src/physics/register_physics.hpp
index 0d9e59bace..8494f4de18 100644
--- a/components/scream/src/physics/register_physics.hpp
+++ b/components/scream/src/physics/register_physics.hpp
@@ -24,6 +24,7 @@ inline void register_physics () {
   proc_factory.register_product("CldFraction",&create_atmosphere_process<CldFraction>);
   proc_factory.register_product("RRTMGP",&create_atmosphere_process<RRTMGPRadiation>);
   proc_factory.register_product("SPA",&create_atmosphere_process<SPA>);
+  proc_factory.register_product("Group",&create_atmosphere_process<AtmosphereProcessGroup>);
 }
 
 } // namespace scream
```

**Comment by AaronDonahue** [comment]
```/usr/tce/packages/gcc/gcc-8.3.1/rh/usr/include/c++/8/ext/aligned_buffer.h(91): error: incomplete type is not allowed
          detected during:
            instantiation of class "__gnu_cxx::__aligned_buffer<_Tp> [with _Tp=scream::AtmosphereProcessGroup]" 
            ```

**Comment by bartgol** [comment]
> ```
> /usr/tce/packages/gcc/gcc-8.3.1/rh/usr/include/c++/8/ext/aligned_buffer.h(91): error: incomplete type is not allowed
>           detected during:
>             instantiation of class "__gnu_cxx::__aligned_buffer<_Tp> [with _Tp=scream::AtmosphereProcessGroup]" 
>             ```
> ```

Well, of course: you need to include the `atmosphere_process_group.hpp` header, otherwise `create_atmosphere_process` does not know how to create one.

**Comment by AaronDonahue** [comment]
Adding where we are at to the git-issue so we don't lose track.  Making this change:
```
]$ git diff
diff --git a/components/scream/src/physics/register_physics.hpp b/components/scream/src/physics/register_physics.hpp
index 0d9e59bace..77b1b4dfda 100644
--- a/components/scream/src/physics/register_physics.hpp
+++ b/components/scream/src/physics/register_physics.hpp
@@ -13,6 +13,7 @@
 #include "physics/cld_fraction/atmosphere_cld_fraction.hpp"
 #include "physics/rrtmgp/atmosphere_radiation.hpp"
 #include "physics/spa/atmosphere_prescribed_aerosol.hpp"
+#include "share/atm_process/atmosphere_process_group.hpp"
 
 namespace scream {
 
@@ -24,6 +25,7 @@ inline void register_physics () {
   proc_factory.register_product("CldFraction",&create_atmosphere_process<CldFraction>);
   proc_factory.register_product("RRTMGP",&create_atmosphere_process<RRTMGPRadiation>);
   proc_factory.register_product("SPA",&create_atmosphere_process<SPA>);
+  proc_factory.register_product("AtmosphereProcessGroup",&create_atmosphere_process<AtmosphereProcessGroup>);
 }
 
 } // namespace scream
```
Gets the code to build.  It still fails at runtime...

**Comment by AaronDonahue** [comment]
Using the following setup in the `input.yaml` file for the stand-alone `shoc_cld_spa_p3_rrtmgp` test
```
Atmosphere Processes:
  Number of Entries: 4
  Schedule Type: Sequential
  Process 0:
    Process Name: SHOC
    Grid: Point Grid
  Process 1:
    Process Name: CldFraction
    Grid: Point Grid
  Process 2:
    Process Name: SPA
    Grid: Point Grid
    SPA Remap File: /p/gpfs1/donahue5/E3SM_simulations/SCREAM/scream_development/sandbox/ctest-build/full_debug/tests/coupled/physics_only/shoc_cld_spa_p3_rrtmgp/data/map_ne4np4_to_ne2np4_mono.nc
    SPA Data File: /p/gpfs1/donahue5/E3SM_simulations/SCREAM/scream_development/sandbox/ctest-build/full_debug/tests/coupled/physics_only/shoc_cld_spa_p3_rrtmgp/data/spa_file_unified_and_complete_ne4_scream.nc
  Process 3:
    Process Name: AtmosphereProcessGroup
    Number of Entries: 2
    Schedule Type: Sequential
    Number of Subcycles: 3
    Process 0:
      Process Name: P3
      Grid: Point Grid
    Process 1:
      Process Name: RRTMGP
      Grid: Point Grid
      active_gases: ["h2o", "co2", "o3", "n2o", "co" , "ch4", "o2", "n2"]
      Orbital Year: 1990
```
fails in P3 (the first subcycled process) suring the initialization of the workspace manager.  See line: `
components/scream/src/physics/p3/atmosphere_microphysics.cpp:278`

**Comment by bartgol** [comment]
This was fixed in #1525.
